src/board/piece.rs

```diff
@@ -0,0 +1,189 @@
+        assert!(legal_rook(0, 56));
+        assert!(legal_rook(56, 0));
+        assert!(legal_rook(8, 15));
+
+        // Same rank
+        assert!(legal_rook(0, 1));
+        assert!(legal_rook(0, 2));
+        assert!(legal_rook(0, 7));
+        assert!(legal_rook(7, 0));
+        assert!(legal_rook(8, 48));
+
+        // Different file && rank
+        assert!(!legal_rook(0, 9));
+        assert!(!legal_rook(0, 17));
+        assert!(!legal_rook(0, 39));
+        assert!(!legal_rook(0, 57));
+        assert!(!legal_rook(0, 63));
+        assert!(!legal_rook(9, 0));
+        assert!(!legal_rook(17, 0));
+        assert!(!legal_rook(39, 0));
+        assert!(!legal_rook(57, 0));
+        assert!(!legal_rook(63, 0));
+
+        // Same square
+        assert!(!legal_rook(0, 0));
+        assert!(!legal_rook(57, 57));
+    }
+
+    #[test]
+    fn test_legal_bishop() {
+        // Diagonal
+        assert!(legal_bishop(35, 26));
+        assert!(legal_bishop(35, 8));
+        assert!(legal_bishop(35, 42));
+        assert!(legal_bishop(35, 56));
+        assert!(legal_bishop(35, 44));
+        assert!(legal_bishop(35, 62));
+        assert!(legal_bishop(35, 28));
+        assert!(legal_bishop(35, 7));
+        assert!(legal_bishop(0, 63));
+        assert!(legal_bishop(63, 0));
+
+        // Horizontal
+        assert!(!legal_bishop(35, 34));
+        assert!(!legal_bishop(35, 32));
+        assert!(!legal_bishop(35, 36));
+        assert!(!legal_bishop(35, 39));
+
+        // Vertical
+        assert!(!legal_bishop(35, 27));
+        assert!(!legal_bishop(35, 3));
+        assert!(!legal_bishop(35, 43));
+        assert!(!legal_bishop(35, 59));
+
+        // Same square
+        assert!(!legal_bishop(35, 35));
+    }
+
+    #[test]
+    fn test_legal_king() {
+        // Move 1 square
+        assert!(legal_king(35, 44)); // ne
+        assert!(legal_king(35, 43)); // up
+        assert!(legal_king(35, 42)); // nw
+        assert!(legal_king(35, 36)); // right
+        assert!(legal_king(35, 34)); // left
+        assert!(legal_king(35, 28)); // se
+        assert!(legal_king(35, 27)); // down
+        assert!(legal_king(35, 26)); // sw
+
+        // Same square
+        assert!(!legal_king(35, 35));
+
+        // Move more than 1 square
+        assert!(!legal_king(35, 47));
+        assert!(!legal_king(35, 51));
+        assert!(!legal_king(35, 45));
+        assert!(!legal_king(35, 37));
+        assert!(!legal_king(35, 33));
+        assert!(!legal_king(35, 25));
+        assert!(!legal_king(35, 19));
+        assert!(!legal_king(35, 22));
+    }
+
+    #[test]
+    fn test_legal_knight() {
+        assert!(legal_knight(0, 17));
+        assert!(legal_knight(0, 15));
+        assert!(legal_knight(0, 10));
+        assert!(legal_knight(0, 6));
+
+        assert!(!legal_knight(0, 16));
+        assert!(!legal_knight(0, 8));
+        assert!(!legal_knight(0, 9));
+        assert!(!legal_knight(0, 7));
+    }
```

Comment:
This doesn't feel like a great way of testing this, but I can't think of anything better.

Reply:
We should be testing through the is_valid_move_for_piece function.